crates/katana/executor/src/abstraction.rs

```diff
@@ -0,0 +1,192 @@
+
+    #[error("Contract ({0}) is not deployed")]
+    ContractNotDeployed(ContractAddress),
+
+    #[error("Invalid transaction nonce. Expected: {expected}, got: {actual}")]
+    InvalidTransactionNonce {
+        /// The actual nonce
+        actual: Nonce,
+        /// The expected nonce
+        expected: Nonce,
+    },
+
+    #[error("Invalid transaction signature")]
+    InvalidSignature,
+
+    #[error(transparent)]
+    Other(Box<dyn std::error::Error + Send>),
+}
+
+pub type ExecutorResult<T> = Result<T, ExecutorError>;
+
+/// Transaction execution simulation flags.
+///
+/// These flags can be used to control the behavior of the transaction execution, such as skipping
+/// the transaction execution or validation, or ignoring the maximum fee when validating the
+/// transaction.
+#[derive(Debug, Clone, Default)]
+pub struct SimulationFlag {
+    /// Skip the transaction execution.
+    pub skip_execute: bool,
+    /// Skip the transaction validation.
+    pub skip_validate: bool,
+    /// Skip checking nonce when validating the transaction.
+    pub skip_nonce_check: bool,
+    /// Skip the fee transfer after the transaction execution.
+    pub skip_fee_transfer: bool,
+    /// Ignore the maximum fee when validating the transaction.
+    pub ignore_max_fee: bool,
+}
+
+/// The output of a executor after a series of executions.
+#[derive(Debug, Default)]
+pub struct ExecutionOutput {
+    /// The state updates produced by the executions.
+    pub states: StateUpdatesWithDeclaredClasses,
+    /// The transactions that have been executed.
+    pub transactions: Vec<(TxWithHash, Option<Receipt>)>,
+}
+
+#[derive(Debug)]
+pub struct EntryPointCall {
+    /// The address of the contract whose function you're calling.
+    pub contract_address: ContractAddress,
+    /// The input to the function.
+    pub calldata: Vec<FieldElement>,
+    /// The contract function name.
+    pub entry_point_selector: FieldElement,
+}
+
+/// A type that can create [BlockExecutor] instance.
+pub trait ExecutorFactory: Send + Sync + 'static {
+    /// Construct a new [BlockExecutor] with the given state.
+    fn with_state<'a, P>(&self, state: P) -> Box<dyn BlockExecutor<'a> + 'a>
+    where
+        P: StateProvider + 'a;
+
+    /// Construct a new [BlockExecutor] with the given state and block environment values.
+    fn with_state_and_block_env<'a, P>(
+        &self,
+        state: P,
+        block_env: BlockEnv,
+    ) -> Box<dyn BlockExecutor<'a> + 'a>
+    where
+        P: StateProvider + 'a;
+
+    /// Returns the configuration environment of the factory.
+    fn cfg(&self) -> &CfgEnv;
+}
+
+/// An executor that can execute a block of transactions.
+pub trait BlockExecutor<'a>: TransactionExecutor + Send + Sync {
+    /// Executes the given block.
+    fn execute_block(&mut self, block: ExecutableBlock) -> ExecutorResult<()>;
+
+    /// Takes the output state of the executor.
+    fn take_execution_output(&mut self) -> ExecutorResult<ExecutionOutput>;
```

Comment:
Cursiosity, why this is not retuned by the execute block function directly?

Reply:
Primarily used by the `IntervalBlockProducer` when [it's mining time](https://github.com/dojoengine/dojo/blob/9823dec0f65743fabf824f6e5e0335ea34bcc580/crates/katana/core/src/service/block_producer.rs#L244) because we execute the transaction as it comes, so we will only need to take the output when the block interval is finished. Compared to instant producer where a transaction can be put in a [block straight away](https://github.com/dojoengine/dojo/blob/9823dec0f65743fabf824f6e5e0335ea34bcc580/crates/katana/core/src/service/block_producer.rs#L425) and be executed as a whole block.